diff --git a/emp-tool/utils/block.h b/emp-tool/utils/block.h
index 45eea46..b124bd6 100644
--- a/emp-tool/utils/block.h
+++ b/emp-tool/utils/block.h
@@ -5,18 +5,6 @@
 #include <immintrin.h>
 #elif __aarch64__
 #include "sse2neon.h"
-inline __m128i _mm_aesimc_si128(__m128i a) {
-	return vreinterpretq_m128i_u8(vaesimcq_u8(vreinterpretq_u8_m128i(a)));
-}
-inline __m128i _mm_aesdec_si128 (__m128i a, __m128i RoundKey)
-{
-    return vreinterpretq_m128i_u8(vaesimcq_u8(vaesdq_u8(vreinterpretq_u8_m128i(a), vdupq_n_u8(0)) ^ vreinterpretq_u8_m128i(RoundKey)));
-}
-
-inline __m128i _mm_aesdeclast_si128 (__m128i a, __m128i RoundKey)
-{
-    return vreinterpretq_m128i_u8(vaesdq_u8(vreinterpretq_u8_m128i(a), vdupq_n_u8(0)) ^ vreinterpretq_u8_m128i(RoundKey));
-}
 #endif
 
 #include <assert.h>
