diff --git a/sources/ippcp/pcpmont_set.c b/sources/ippcp/pcpmont_set.c
index 00adfeb6..3b4907ff 100644
--- a/sources/ippcp/pcpmont_set.c
+++ b/sources/ippcp/pcpmont_set.c
@@ -35,7 +35,7 @@ __INLINE int cpGetBitSize(Ipp32u offset, Ipp32u val)
 {
     int bitSize = 32;
     if (val == 0) return 0;
-    while ((val & (1 << bitSize)) == 0) bitSize--;
+    while ((val & (1 << (bitSize - 1))) == 0) bitSize--;
     return (int)offset + bitSize;
 }
 