diff --git a/CMakeLists.txt b/CMakeLists.txt
index dfbcb0f..72f9a46 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -113,7 +113,7 @@ set_target_properties(${PROJECT_NAME} PROPERTIES
     OUTPUT_NAME tommath
     VERSION ${PROJECT_VERSION}
     SOVERSION ${PROJECT_VERSION_MAJOR}
-    PUBLIC_HEADER "${PUBLIC_HEADERS}"
+    PUBLIC_HEADER "${HEADERS}"
 )

 option(COMPILE_LTO "Build with LTO enabled")
diff --git a/tommath_private.h b/tommath_private.h
index d319a1d..5f4446e 100644
--- a/tommath_private.h
+++ b/tommath_private.h
@@ -17,7 +17,7 @@
  * On Win32 a .def file must be used to specify the exported symbols.
  */
 #if defined(__GNUC__) && __GNUC__ >= 4 && !defined(_WIN32) && !defined(__CYGWIN__)
-#   define MP_PRIVATE __attribute__ ((visibility ("hidden")))
+#   define MP_PRIVATE
 #else
 #   define MP_PRIVATE
 #endif
